examples/discover/main.go

```diff
@@ -0,0 +1,319 @@
+	http.Handle("localhost/oras/artifacts/v1/net-monitor/manifests/",
+		&mockRegistry{
+			ref:         "localhost:5000/net-monitor:v1",
+			memorystore: memoryStore,
+			references:  example_references})
+
+	go http.ListenAndServe(":5000", http.DefaultServeMux)
+
+	localhostResolver = memoryStore
+}
+
+func init() {
+	target := content.NewMemory()
+
+	targetMux := &http.ServeMux{}
+	targetMux.Handle("localhost/v2/net-monitor/manifests/v1",
+		&mockRegistry{
+			ref:         "localhost:5002/net-monitor:v1",
+			memorystore: target})
+	targetMux.Handle("localhost/oras/artifacts/v1/net-monitor/manifests",
+		&mockRegistry{
+			ref:         "localhost:5002/net-monitor:v1",
+			memorystore: target,
+			references:  example_references,
+		})
+
+	go http.ListenAndServe(":5002", targetMux)
+
+	targetResolver = target
+}
+
+type mockRegistry struct {
+	ref         string
+	memorystore *content.Memory
+	references  []artifactspec.Descriptor
+}
+
+func (e *mockRegistry) ServeHTTP(writer http.ResponseWriter, req *http.Request) {
+	if strings.Contains(req.URL.String(), "blobs") {
+		return
+	}
+
+	_, desc, err := e.memorystore.Resolve(req.Context(), e.ref)
+	if err != nil {
+		writer.WriteHeader(http.StatusNotFound)
+		return
+	}
+
+	if req.Method == http.MethodHead {
+		headers := writer.Header()
+		headers.Set("Docker-Content-Digest", desc.Digest.String())
+		headers.Set("Content-Type", desc.MediaType)
+		headers.Set("Content-Length", fmt.Sprint(desc.Size))
+		writer.WriteHeader(http.StatusOK)
+		return
+	}
+
+	if req.Method != http.MethodGet {
+		writer.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	if strings.Contains(req.URL.Path, "referrers") && strings.Contains(req.URL.Path, "sha256:d9cbd5e9195d3d68fd126ec8fcd687042d1eb767c57de4d776830c3545b534d4") {
+		exampleArtifacts := struct {
+			References []artifactspec.Descriptor `json:"references"`
+		}{
+			References: e.references,
+		}
+
+		exampleArtifacts.References[0].Annotations["test-filter"] = "tested"
+
+		c, err := json.Marshal(exampleArtifacts)
+		if err != nil {
+			writer.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+
+		written, err := writer.Write(c)
+		if err != nil {
+			writer.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+
+		writer.Header().Set("Docker-Content-Digest", string(desc.Digest))
+		writer.Header().Set("Content-Length", fmt.Sprint(written))
+		writer.Header().Set("Content-Type", "application/json")
+		return
+	}
+
+	desc, content, ok := e.memorystore.Get(desc)
+	if !ok {
+		writer.WriteHeader(http.StatusNoContent)
+		return
+	}
+
+	written, err := writer.Write(content)
+	if err != nil {
+		writer.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	if written <= 0 {
+		writer.WriteHeader(http.StatusAccepted)
+		return
+	}
+
+	writer.Header().Set("Docker-Content-Digest", string(desc.Digest))
+	writer.Header().Set("Content-Type", desc.MediaType)
+	writer.Header().Set("Content-Length", fmt.Sprint(desc.Size))
+}
```

Comment:
Since it is an example not an unit test, it is better to assume that there is already a local registry running in the background. Meanwhile, demostrate how to push artifact and artfact manifest. Then demotestate how to discover and pull everything down.

Reply:
Pushing artifacts is already covered in other examples, this is just about discovery.